### PR TITLE
Note tags

### PR DESCRIPTION
**Pull request by anitaschiller**
Branch for user stories: Create remember tags, save and delete single notes in header

The remember section of each entry now consists of tags instead of static notes. After submission of the form, single tags can be saved and shown in the header of each details page. The same works reverse and tags can be unsaved on click.

Next step: Add groups

Thank you!! :-)

### CODE REVIEW

File: client/src/components/EntryCard.js [hunk 1]
```diff
@@ -1,18 +1,29 @@
 import styled from 'styled-components';
+import { StarIconEmpty } from '../icons/StarIconEmpty';
+import { StarIconFilled } from '../icons/StarIconFilled';
 
-export default function EntryCard({ entry, onDeleteEntry }) {
+export default function EntryCard({ entry, onDeleteEntry, onToggleNote }) {
   return (
     <CardWrapper>
       <DeleteIcon onClick={() => onDeleteEntry(entry.id)}>x</DeleteIcon>
```
Comment (jamarob) [inline client/src/components/EntryCard.js]: Some people prefer `&times;` for the closing x as it is not distorted by the used font.

File: client/src/components/EntryCard.js [hunk 2]
```diff
@@ -1,18 +1,29 @@
...
       <Headline>
         {entry.date} · {entry.title}
       </Headline>
-      <p>{entry.remember}</p>
+      {entry.remember.map((note) => (
+        <NoteTag onClick={() => onToggleNote(note)} key={note.id}>
+          {note.isSaved ? <StarFilledStyled /> : <StarEmptyStyled />}
+          <Note>{note.noteTag}</Note>
+        </NoteTag>
+      ))}
```
Comment (jamarob) [inline client/src/components/EntryCard.js]: Imho the naming is a little bit confusing
- you map over `entry.remember` but the items are called `note`, so why is this named `remember` and not `notes`?
- `NoteTag` has `Note` as a child which has `note.noteTag` as content 🤔 

Maybe reserve some time for a refactoring ticket to improve in this.

File: client/src/components/NoteTags.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
+import styled from 'styled-components';
```
Comment (jamarob) [inline client/src/components/NoteTags.js]: Maybe use `styled-components/macro`, it's great for debugging!

File: client/src/components/NoteTags.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    setTagValue(event.target.value);
+  }
+
+  function handleKeyDown(event) {
+    if (event.key === 'Enter') {
+      event.preventDefault();
+      const memberTags = [...tags, tagValue];
+      setTags(memberTags);
+      onCreateTag(tagValue);
+      setTagValue([]);
```
Comment (jamarob) [inline client/src/components/NoteTags.js]:
```suggestion
      setTagValue('');
```
Why an array, the initial state is an empty string?

File: client/src/components/NoteTags.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  function handleChange(event) {
+    setTagValue(event.target.value);
+  }
+
+  function handleKeyDown(event) {
+    if (event.key === 'Enter') {
+      event.preventDefault();
+      const memberTags = [...tags, tagValue];
+      setTags(memberTags);
+      onCreateTag(tagValue);
```
Comment (jamarob) [inline client/src/components/NoteTags.js]: Maybe rethink this, you pass down ´setTags´ as a prop but can call it _inside_ of `onCreateTag` in the component above.

File: client/src/components/NoteTags.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+      setTagValue([]);
+    }
+  }
+
+  return (
+    <>
+      <label htmlFor="remember">Remember</label>
```
Comment (jamarob) [inline client/src/components/NoteTags.js:22]: `htmlFor` works with ids, which component is this referencing?

File: client/src/components/NoteTags.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+const TagsWrapper = styled.div`
+  display: flex;
+  flex-direction: row;
+  flex-wrap: wrap;
+
+  background: white;
+  border: 1px solid #a8a8a8;
+  border-radius: 5px;
+  height: fit-content;
```
Comment (jamarob) [inline client/src/components/NoteTags.js:50]: I think this can be deleted, no sure..

File: client/src/components/NoteTags.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+const Tag = styled.span`
+  background: var(--secondary);
+  color: white;
+  margin: 0.1rem;
+  padding: 0.3rem;
+`;
+
+const TagInput = styled.input`
+  border: none;
+  outline: none;
```
Comment (jamarob) [inline client/src/components/NoteTags.js:63]: Maybe provide an alternative to outline using ´:focus´ to signal that the element is active to the user?

File: client/src/icons/StarIconEmpty.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+const Svg = styled(Icon)`
+  width: 31px;
+  height: 30px;
+`;
+
+export const StarIconEmpty = ({ className }) => (
+  <Svg viewBox="0 0 30.849 29.339" className={className}>
+    <path
+      fill="currentColor"
+      data-name="Icon metro-star-empty"
```
Comment (jamarob) [inline client/src/icons/StarIconEmpty.js]:
```suggestion
      dataName="Icon metro-star-empty"
```

File: client/src/icons/StarIconFilled.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+const Svg = styled(Icon)`
+  width: 31px;
+  height: 30px;
+`;
+
+export const StarIconFilled = ({ className }) => (
+  <Svg viewBox="0 0 30.849 29.339" className={className}>
+    <path
+      fill="currentColor"
+      data-name="Icon metro-star-empty"
```
Comment (jamarob) [inline client/src/icons/StarIconFilled.js]:
```suggestion
      dataName="Icon metro-star-empty"
```

File: client/src/pages/Details.js [hunk 11]
```diff
@@ -1,26 +1,51 @@
...
+import { StarIconFilled } from '../icons/StarIconFilled';
 import { UnfoldIcon } from '../icons/UnfoldIcon';
 
 export default function Details({ member, updateMember, members }) {
   const initialEntry = {
     date: '',
     title: '',
-    remember: '',
+    remember: [],
   };
```
Comment (jamarob) [inline client/src/pages/Details.js:17]: This can be moved outside of the function.

File: client/src/pages/Details.js [hunk 12]
```diff
@@ -1,26 +1,51 @@
...
+      entryRemember.map((entry) =>
+        entry.forEach((note) => {
+          allNotes.push(note);
+        })
+      );
+
+      const currentlySavedNotes = allNotes.filter(
+        (note) => note.isSaved === true
+      );
+      return currentlySavedNotes;
```
Comment (jamarob) [inline client/src/pages/Details.js]:
```suggestion
return newMember.entries
              .flatMap(entry => entry.remember)
              .filter(note => note.isSaved)
```

File: client/src/pages/Details.js [hunk 13]
```diff
@@ -1,26 +1,51 @@
...
   const [newMember, setNewMember] = useState(
     members.find((newMember) => newMember.id === member.id)
   );
-
   const [entry, setEntry] = useState(initialEntry);
   const [entries, setEntries] = useState(newMember.entries ?? []);
   const [isError, setIsError] = useState(false);
   const [isUnfolded, setIsUnfolded] = useState(false);
+  const [tags, setTags] = useState([]);
+  const [savedNotes, setSavedNotes] = useState(findSavedNotes() ?? []);
```
Comment (jamarob) [inline client/src/pages/Details.js]: I think you can simplify this a little, you don't need all these states for derived values, simply use a const that will get reevaluated every time the component rerenders.

Think about `newMember`: it is derived from `member`, when you call `updateMember` the state higher up changes, the `members` prop changes in turn and the component rerenders - no need for an extra state.

I think you don't need `member` as a prop anyway, you only need the id, which will probably not change and thus prevent further unnecessary rerendering.

File: client/src/pages/Details.js [hunk 14]
```diff
@@ -70,6 +137,15 @@ export default function Details({ member, updateMember, members }) {
         </DetailsHeadline>
         <DetailsGroup>{newMember.group}</DetailsGroup>
         <p>{newMember.description}</p>
+        <SavedNoteWrapper>
+          {savedNotes.length >= 1 &&
```
Comment (jamarob) [inline client/src/pages/Details.js]: `savedNotes` will always be an array so it is save to map over it without the guard

File: client/src/pages/Details.js [hunk 15]
```diff
@@ -78,25 +154,25 @@ export default function Details({ member, updateMember, members }) {
         {isUnfolded && (
           <>
             <label htmlFor="date">Date</label>
```
Comment (jamarob) [inline client/src/pages/Details.js:127]: The following input needs to have an id for `htmlFor` to work

File: client/src/pages/Details.js [hunk 16]
```diff
@@ -78,25 +154,25 @@ export default function Details({ member, updateMember, members }) {
...
-            <input
+            <Input
               type="text"
               name="date"
               value={entry.date}
               onChange={changeHandler}
             />
             <label htmlFor="title">Title</label>
```
Comment (jamarob) [inline client/src/pages/Details.js:135]: The following input needs to have an id for `htmlFor` to work

File: client/src/pages/Details.js [hunk 17]
```diff
@@ -166,20 +243,45 @@ const FormStyled = styled.form`
...
+const SavedNote = styled.span`
+  border: solid #a8a8a8 1px;
+  border-radius: 5px;
+  margin: 0 0.4rem 0.4rem 0;
+  padding: 0.5rem;
+  position: relative;
+`;
+
+const SavedNoteWrapper = styled.div`
+  margin: 1.5rem 0rem 0.8rem 0;
```
Comment (jamarob) [inline client/src/pages/Details.js]: 0 without unit

**Review by jamarob — COMMENTED** [review]
Good job! 

Left some comments and suggest you refactor the `Details` component. I think you can get rid of some of the states, also think about writing a custom hook to encapsulated all the logic. ✌️

File: client/src/components/EntryCard.js (hunk 2 above)
Comment (anitaschiller) [reply client/src/components/EntryCard.js]: You are right, the naming doesn't make sense. I changed noteTag to noteContent! :-)

File: client/src/components/NoteTags.js (hunk 4 above)
Comment (anitaschiller) [reply client/src/components/NoteTags.js]: Ohh yes, thanks for checking!

File: client/src/pages/Details.js (hunk 15 above)
Comment (anitaschiller) [reply client/src/pages/Details.js:127]: Thank you!!

**Comment by anitaschiller** [comment]
> Good job!
> 
> Left some comments and suggest you refactor the `Details` component. I think you can get rid of some of the states, also think about writing a custom hook to encapsulated all the logic. ✌️

Thank you for all your comments, they were really helpful! 😃

File: client/src/components/NoteTags.js (hunk 6 above)
Comment (anitaschiller) [reply client/src/components/NoteTags.js:22]: It's part of the form on the details page. I added all IDs accordingly.

File: client/src/pages/Details.js (hunk 12 above)
Comment (anitaschiller) [reply client/src/pages/Details.js]: Wow, that's really much much easier 😅

File: client/src/pages/Details.js (hunk 13 above)
Comment (anitaschiller) [reply client/src/pages/Details.js]: I had to take a deeper look first, but you are right, rerendering of newMember is not needed! Thank you :-)

**Pull request merged**